lib/src/value/color.ts

```diff
@@ -3,202 +3,372 @@
-    if (this.saturationInternal === undefined) {
-      this.rgbToHsl();
+    emitColor4ApiDeprecation('saturation');
+    try {
+      return get(this.color, 'saturation');
+    } catch (error) {
+      return get(to(this.color, 'hsl'), 'saturation');
     }
-    return this.saturationInternal!;
   }
 
   /** `this`'s hue value. */
   get lightness(): number {
-    if (this.lightnessInternal === undefined) {
-      this.rgbToHsl();
+    emitColor4ApiDeprecation('lightness');
+    try {
+      return get(this.color, 'lightness');
+    } catch (error) {
+      return get(to(this.color, 'hsl'), 'lightness');
     }
-    return this.lightnessInternal!;
   }
 
   /** `this`'s whiteness value. */
   get whiteness(): number {
-    // Because HWB is (currently) used much less frequently than HSL or RGB, we
-    // don't cache its values because we expect the memory overhead of doing so
-    // to outweigh the cost of recalculating it on access.
-    return (Math.min(this.red, this.green, this.blue) / 255) * 100;
+    emitColor4ApiDeprecation('whiteness');
+    try {
+      return get(this.color, 'whiteness');
+    } catch (error) {
+      return get(to(this.color, 'hwb'), 'whiteness');
+    }
   }
 
   /** `this`'s blackness value. */
   get blackness(): number {
-    // Because HWB is (currently) used much less frequently than HSL or RGB, we
-    // don't cache its values because we expect the memory overhead of doing so
-    // to outweigh the cost of recalculating it on access.
-    return 100 - (Math.max(this.red, this.green, this.blue) / 255) * 100;
+    emitColor4ApiDeprecation('blackness');
+    try {
+      return get(this.color, 'blackness');
+    } catch (error) {
+      return get(to(this.color, 'hwb'), 'blackness');
+    }
   }
 
   /** `this`'s alpha channel. */
   get alpha(): number {
-    return this.alphaInternal;
+    return this.color.alpha;
   }
 
-  /**
-   * Whether `this` has already calculated the HSL components for the color.
+  /** `this`'s color space. */
+  get space(): string {
+    const _space = this.color.space.id;
+    if (_space === 'srgb' && this.format === 'rgb') {
+      return 'rgb';
+    }
+    return _space;
+  }
+
+  /** Whether `this` is in a legacy color space. */
+  get isLegacy(): boolean {
+    return ['rgb', 'hsl', 'hwb'].includes(this.space);
+  }
+
+  /** The values of this color's channels (excluding the alpha channel), or
+   * `null` for [missing] channels.
    *
-   * This is an internal property that's not an official part of Sass's JS API,
-   * and may be broken at any time.
+   * [missing]: https://www.w3.org/TR/css-color-4/#missing
    */
-  get hasCalculatedHsl(): boolean {
-    return !!this.hueInternal;
+  get channelsOrNull(): List<number | null> {
+    return List(this.color.coords.map(c => (Number.isNaN(c) ? null : c)));
+  }
+
+  /** The values of this color's channels (excluding the alpha channel). */
+  get channels(): List<number> {
+    return List(this.color.coords);
```

Comment:
This needs to convert NaNs to 0. Probably most methods here do. Definitely something to keep an eye out for in the API tests.

Reply:
Adjusted in https://github.com/sass/embedded-host-node/pull/259/commits/04010b6960e5c45fc3f5859655a79f91da3efc4c